Issue 2:

apavlov1992 opened the pull request:
- Make float project structure;
- Add config.yaml with default parameters to parse;
- Add new arguments to main.go as n string - Number of comics, and bool o - output final JSON or not;
- Add build stages in Makefile.

Commit:
Issue 2:
- Make float project structure;
- Add config.yaml with default parameters to parse;
- Add new arguments to main.go as n string - Number of comics, and bool o - output final JSON or not;
- Add build stages in Makefile.

Commit:
Issue 2:
- Make float project structure;
- Add config.yaml with default parameters to parse;
- Add new arguments to main.go as n string - Number of comics, and bool o - output final JSON or not;
- Add build stages in Makefile.

Commit:
Issue 2:
- Make float project structure;
- Add config.yaml with default parameters to parse;
- Add new arguments to main.go as n string - Number of comics, and bool o - output final JSON or not;
- Add build stages in Makefile.

vbelogrudov commented on cmd/main.go:
```diff
@@ -0,0 +1,44 @@
+func main() {
+	var configFileName string
+	flag.StringVar(&configFileName, "config", "../config/config.yaml", "Specify configuration file name to use.")
+	numberOfComics := flag.Int("n", 10, "Number of comics to fetch")
```
у нас было number of comics to print ;) Fetch all ...

apavlov1992 replied:
Так принтим, если флаг "o"

vbelogrudov commented on cmd/main.go:
```diff
@@ -0,0 +1,44 @@
+	}
+
+	client := xkcd.Client{
+		Client:  http.Client{},
```
пустой клиент не отличается от дефолтного, который можно звать через http.Get .. Сюда можно хотя бы Timeout секунд 10 вставить в параметры клиента

apavlov1992 replied:
Добавил timeout 5 секунд

vbelogrudov commented on internal/stemming/stemming.go:
```diff
@@ -37,11 +37,12 @@ func DownloadFile(filepath string, url string) error {
 
 func isStopWord(word string) bool {
+
 	return slices.Contains(stopWordList, word)
```
в идеале должен быть set (map[string]bool), иначе мы много будем бегать по списку

apavlov1992 replied:
переделал

vbelogrudov commented on internal/xkcd/xkcd.go:
```diff
@@ -0,0 +1,63 @@
+}
+
+type ComicsInfo struct {
+	Num         interface{}
```
interface -> int или string. Если не хотим печатать в json - `json:"-"`

apavlov1992 replied:
Переделал на int

vbelogrudov commented on internal/xkcd/xkcd.go:
```diff
@@ -0,0 +1,63 @@
+		if err != nil {
+			log.Fatal(err)
+		}
+
```
здесь не хватает defer resp.Body.Close()

apavlov1992 replied:
Добавил

vbelogrudov commented on internal/xkcd/xkcd.go:
```diff
@@ -0,0 +1,63 @@
+			log.Fatal(err)
+		}
+
+		bodyBytes, err := io.ReadAll(resp.Body)
```
можно err = json.NewDecoder(resp.Body).Decode(&comics)

apavlov1992 replied:
done

vbelogrudov commented on internal/xkcd/xkcd.go:
```diff
@@ -0,0 +1,63 @@
+		if err != nil {
+			log.Fatal(err)
+		}
+		comic.Description = strings.Join(
```
стемминг лучше делать в main или любом другом связующем пакете, тогда xkcd не зависит от stemming

apavlov1992 replied:
перевёл стемминг в main, убрал его из других пакетов

vbelogrudov commented on Makefile:
```diff
@@ -1,3 +1,8 @@
+.PHONY: build/xkcd build/stemming
 
+build/xkcd: ## Build binary file for xkcd
+	env GOOS=linux GOARCH=amd64 CGO_ENABLED=0 go build -o xkcd internal/xkcd/xkcd.go
```
нужен файл  с main

apavlov1992 replied:
done

vbelogrudov replied:
not yet :)

vbelogrudov replied:
make -> xkcd -> permission denied (not executable)

vbelogrudov reviewed (approved):
небольшие правки, еще бы сюда определение максимального id комикса, лежит в
xkcd.com/info.0.json

Commit:
Issue 2:
- Make float project structure;
- Add config.yaml with default parameters to parse;
- Add new arguments to main.go as n string - Number of comics, and bool o - output final JSON or not;
- Add build stages in Makefile;
- Fix PR issues.

Commit:
Issue 2:
- Make float project structure;
- Add config.yaml with default parameters to parse;
- Add new arguments to main.go as n string - Number of comics, and bool o - output final JSON or not;
- Add build stages in Makefile;
- Fix PR issues.

apavlov1992 commented:
> небольшие правки, еще бы сюда определение максимального id комикса, лежит в xkcd.com/info.0.json

доделал

vbelogrudov commented:
не компилируется ;)

vbelogrudov commented on cmd/main.go:
```diff
@@ -0,0 +1,57 @@
+func main() {
+	var configFileName string
+	flag.StringVar(&configFileName, "config", "../config/config.yaml", "Specify configuration file name to use.")
+	numberOfComics := flag.Int("n", 0, "Number of comics to fetch")
```
IntVar, чтобы потом не возиться с поинтерами

vbelogrudov commented on cmd/main.go:
```diff
@@ -0,0 +1,57 @@
+
+	if *numberOfComics == 0 {
+		countOfAllComics := client.GetMaxId()
+		numberOfComics = &countOfAllComics
```
это выглядит по-хакерски, лучше *numberOfComics = countOfAllComics, а лучше вообще без указателей

vbelogrudov commented on internal/xkcd/xkcd.go:
```diff
@@ -0,0 +1,68 @@
+	}
+
+	defer resp.Body.Close()
+	err = json.NewDecoder(resp.Body).Decode(&comic)
```
не хватает обработки ошибки

vbelogrudov commented on internal/stemming/stemming.go:
```diff
@@ -37,11 +38,12 @@ func DownloadFile(filepath string, url string) error {
-func StemmingWords(words []string) string {
-	wordsList := []string{} //to do map
+func stemmingWords(words []string) []string {
+	wordsList := []string{}
```
var wordList []string // неинициализированный слайс лучше, чем пустой

vbelogrudov commented on internal/stemming/stemming.go:
```diff
@@ -50,28 +52,27 @@ func StemmingWords(words []string) string {
 	stopWordList = strings.Split(string(fileData), ",")
+	wordsMap = make(map[string]any)
+	for _, word := range stopWordList {
+		wordsMap[word] = struct{}{}
```
wordsMap = make(map[string]bool)
...
     wordsMap[word] = true

А когда проверяем, достаточно if wordsMap[word], без ok

Commit:
Issue 2:
- Make float project structure;
- Add config.yaml with default parameters to parse;
- Add new arguments to main.go as n string - Number of comics, and bool o - output final JSON or not;
- Add build stages in Makefile;
- Fix PR issues.
